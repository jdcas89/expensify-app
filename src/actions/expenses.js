import uuid from "uuid";
import * as types from './actionTypes';
import database from '../firebase/firebase';




export const addExpense = (expense) => ({ 
  type: types.ADD_EXPENSE, 
  expense
});

export const removeExpense = ({ id }) => ({
  type: types.REMOVE_EXPENSE,
  id
});

export const editExpense = (id, updates) => ({
  type: types.EDIT_EXPENSE,
  id,
  updates
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "", note = "", amount = 0, createdAt = 0
    } = expenseData;
    const expense = {description, note, amount , createdAt}
    database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}