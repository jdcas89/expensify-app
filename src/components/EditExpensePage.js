import React from "react";

const EditExpensePage = (props) => {
  console.log(props)
  return (
    <div>
      <h2>Edit Page</h2>
      <p>Editing the page of item {props.match.params.id}</p>
    </div>
  );
};

export default EditExpensePage;
