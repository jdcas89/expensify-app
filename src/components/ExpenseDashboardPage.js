import React from 'react';

const ExpenseDashboardPage = () => (
  <div>
    <h2>Home Page</h2>
  </div>
);

export default ExpenseDashboardPage;