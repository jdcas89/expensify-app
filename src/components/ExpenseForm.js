import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      errorMessage: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    e.persist();
    this.setState(() => ({ note: e.target.value }));
  };

  // https://regex101.com/
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        errorMessage: "You have to provide a description and an amount!"
      }));
    } else {
      this.setState(() => ({ errorMessage: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.errorMessage && <p>{this.state.errorMessage}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            placeholder="Description"
            autoFocus
          />
          <input
            type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder="Amount"
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            value={this.state.note}
            onChange={this.onNoteChange}
            placeholder="Add a note for your expense"
          />

          <button>Add expense</button>
        </form>
      </div>
    );
  }
}
