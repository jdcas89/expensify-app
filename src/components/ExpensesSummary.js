import React from "react";
import { connect } from "react-redux";
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from "../selectors/expenses";
import numeral from 'numeral';

export const ExpensesSummary = ({expenseCount, expenseTotal}) => (
  <div>
  {
    expenseCount === 0 ? (<p>There are no expenses.</p>) : 
    (
      expenseCount === 1 ? (<p>Viewing {expenseCount} expense totalling {numeral(expenseTotal/100).format("$0,0.00")}</p>) 
      : 
      (
        <p>Viewing {expenseCount} expenses totalling {numeral(expenseTotal/100).format("$0,0.00")}</p>
      )
    )
  }
  </div>
)

const mapStateToProps = (state, props) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)
  return {
  expenseCount: visibleExpenses.length,
  expenseTotal: selectExpensesTotal(visibleExpenses)
}
};

export default connect(mapStateToProps)(ExpensesSummary);
