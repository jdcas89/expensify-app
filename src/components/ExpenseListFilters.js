import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";


class ExpenseListFilters extends React.Component {
  constructor(props){
    super(props);
    this.state = {
      calendarFocused: null
    }
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({calendarFocused}))
  }
  onDatesChange = ({startDate, endDate}) => {
    this.props.dispatch(setStartDate(startDate))
    this.props.dispatch(setEndDate(endDate))
  }
  render(){
    return (
      <div>
        <p>Filter text: </p>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => {
            props.dispatch(setTextFilter(e.target.value));
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={(e) => {
            console.log(e.target.value)
            if(e.target.value === 'date'){
              this.props.dispatch(sortByDate());
            }else if(e.target.value === 'amount'){
              this.props.dispatch(sortByAmount());
            }
          }}
          >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
        startDate={this.props.filters.startDate} 
        endDate={this.props.filters.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        numberOfMonths={1}
        isOutsideRange= {() => false}
        showClearDates={true}
      />      
      </div>
    )
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
