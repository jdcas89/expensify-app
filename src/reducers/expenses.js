import * as types from '../actions/actionTypes'
//Expenses reducer

const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case types.ADD_EXPENSE:
      return [...state, action.expense]
    case types.EDIT_EXPENSE:
      return state.map((expense) => {
        if(expense.id === action.id){
          return {
            ...expense,
            ...action.updates
          };
        }else{
          return expense;
        }
      })
    case types.REMOVE_EXPENSE:
      return state.filter(({ id }) => id !== action.id);
    case types.SET_EXPENSES:
      return action.expenses;
    default:
      return state;
  }
};

//export default expensesReducer;