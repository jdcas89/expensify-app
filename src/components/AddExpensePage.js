import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  addExpense = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h2>Create Expense</h2>
        <ExpenseForm onSubmit={this.addExpense} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
