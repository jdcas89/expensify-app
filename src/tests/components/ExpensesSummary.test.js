import React from 'react';
import { shallow }  from 'enzyme';
import  { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';


test('should render ExpensesSummary correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={255}/>)
  expect(wrapper).toMatchSnapshot();
});


test('should render a paragraph if no expenses are loaded', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={0} expenseTotal={0}/>)
  expect(wrapper.find('p').text()).toBe('There are no expenses.')
});

test('should render a paragraph if one expense is loaded', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={795}/>)
  expect(wrapper.find('p').text()).toBe('Viewing 1 expense totalling $7.95')
});

test('should render a paragraph if expenses are loaded', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={5} expenseTotal={5500}/>)
  
  expect(wrapper.find('p').text()).toBe('Viewing 5 expenses totalling $55.00')
});
