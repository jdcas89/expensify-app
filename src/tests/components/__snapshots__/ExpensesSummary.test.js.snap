// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     expense totalling 
    $2.55
  </p>
</div>
`;
